src/future/flatten_sink.rs

```diff
@@ -0,0 +1,91 @@
+use std::mem;
+
+use {Async, AsyncSink, StartSend, Poll, Future, Sink};
+
+
+enum State<F: Future>
+    where F::Item: Sink
+{
+    Future(F),
+    Sink(F::Item),
+    Error(<F::Item as Sink>::SinkError),
+    Void,
+}
+
+/// Future for the `into_sink` combinator, flattening a future-of-a-sink to
+/// get just the result of the final sink instead
+///
+/// This is created by the `Future::into_sink` method.
+pub struct FlattenSink<F: Future>
+    where F::Item: Sink
+{
+    state: State<F>,
+}
+
+pub fn new<F>(future: F) -> FlattenSink<F>
+    where F: Future,
+          F::Item: Sink
+{
+    FlattenSink {
+        state: State::Future(future),
+    }
+}
+
+impl<F: Future> Sink for FlattenSink<F>
+    where F::Item: Sink,
+          <F::Item as Sink>::SinkError: From<F::Error>,
+{
+    type SinkItem = <F::Item as Sink>::SinkItem;
+    type SinkError = <F::Item as Sink>::SinkError;
+    fn start_send(&mut self, item: Self::SinkItem)
+        -> StartSend<Self::SinkItem, Self::SinkError>
+    {
+        let (res, state) = match mem::replace(&mut self.state, State::Void) {
+            State::Future(mut fut) => {
+                match fut.poll() {
+                    Ok(Async::Ready(mut sink)) => {
+                        (sink.start_send(item)?, State::Sink(sink))
```

Comment:
> I figure that if there's interest in differentiating the future error from the sink error this combinator wouldn't be used, right? Instead it'd be handled manually by having the error handling for the future be slightly different?

Yes, good point. But this only proves what I'm saying. Let me elaborate on it:

1. Let's say we have a `Sink` that returns `io::Error`. And we know that if it returns `io::ErrorKind::Other` it is resumable.
2. Then we use `flatten_stream` with arbitrary future that also returns `io::Error`, errors are compatible, so we're happy to use it.
3. Then on some very rare occasion Future returns `Other` too and we know that we could resume when `Other` error is returned. So we restart and get panic.

This kind of situation is what I mean by this point:

>  user still has to differentiate the error generated by Future and error generated by Sink

So the only safe way to make `flatten_stream` resumable on error is to return `enum Error { FutureError(A), SinkError(B) }`, which is too narrow use case to include here I think.